Doc/library/collections.rst

```diff
@@ -1252,16 +1252,17 @@ variants of :func:`functools.lru_cache`:
                 self.cache.move_to_end(args)
                 return self.cache[args]
             result = self.func(*args)
-            self.requests[args] = self.requests.get(args, 0) + 1
+            self.requests[args] = self.requests.get(args, 0) + 1 # increment existing value by 1 if seen before, otherwise set to 1
             if self.requests[args] <= self.cache_after:
                 self.requests.move_to_end(args)
                 if len(self.requests) > self.maxrequests:
-                    self.requests.popitem(0)
+                    self.requests.popitem(last=False)
             else:
-                self.requests.pop(args, None)
+                # entry has been seen more than cache_after times
+                self.requests.pop(args) # no longer need to keep track of how many times this entry has been seen
+                if len(self.cache) == self.maxsize:
```

Comment:
I do not agree with this code edit.  Letting the cache grow larger than the maxsize and then popping is safer because we are guaranteed that at least one entry is present. Also, we tend to put deletion code after the insertion code so that the cache is in a consistent state prior to deletion whose DECREF logic can trigger arbitrary Python code.  

Please restore the original code order.

Reply:
I reverted this change in commit 26f9f9061d4b196845fd82da3738ccddcfa7ab7c.

I think I got the gist of what you're saying here, but I don't think I understand the internals of CPython enough to follow all of it.

> Letting the cache grow larger than the maxsize and then popping is safer because we are guaranteed that at least one entry is present.

Is this also to avoid issues in multi-threaded environments, as you mentioned above (and in issue 107421)?

> Also, we tend to put deletion code after the insertion code so that the cache is in a consistent state prior to deletion whose DECREF logic can trigger arbitrary Python code.

I think this is referring to garbage collection that might remove the cache when the cache becomes empty, correct? In that case, wouldn't the `self.cache` still be safe from garbage collection since the reference count is still (at least) 1 since the instance has a reference to the `self.cache` attribute? Or is this implementation dependent? (Or is it something else altogether? 😅)